src/com/linkedin/parseq/internal/SerialExecutor.java

```diff
@@ -69,24 +70,40 @@
   private final Executor _executor;
   private final RejectedSerialExecutionHandler _rejectionHandler;
   private final ExecutorLoop _executorLoop = new ExecutorLoop();
-  private final FIFOPriorityQueue<Runnable> _queue = new FIFOPriorityQueue<Runnable>();
+  private final TaskQueue<PrioritizableRunnable> _queue;
   private final AtomicInteger _pendingCount = new AtomicInteger();
   private final DeactivationListener _deactivationListener;
 
   public SerialExecutor(final Executor executor,
       final RejectedSerialExecutionHandler rejectionHandler,
-      final DeactivationListener deactivationListener) {
+      final DeactivationListener deactivationListener,
+      final TaskQueue<PrioritizableRunnable> taskQueue) {
     ArgumentUtil.requireNotNull(executor, "executor");
     ArgumentUtil.requireNotNull(rejectionHandler, "rejectionHandler" );
     ArgumentUtil.requireNotNull(deactivationListener, "deactivationListener" );
 
     _executor = executor;
     _rejectionHandler = rejectionHandler;
+    _queue = taskQueue;
     _deactivationListener = deactivationListener;
   }
 
   public void execute(final Runnable runnable) {
-    _queue.add(runnable);
+    if (runnable instanceof Prioritizable) {
+      _queue.add((PrioritizableRunnable)runnable);
+    } else {
+      _queue.add(new PrioritizableRunnable() {
```

Comment:
We could. But it means we'll have to do an unchecked type casting which can fail hard. Alternatively, we can drop the "implements Executor" declaration on SerialExecutor and enforce PrioritizableRunnable at method level. Note that SerialExecutor is already pretty tightly coupled with engine because we let user wire in PlanDeactivationListener and TaskQueue, so I don't see much value to keep the Executor interface.